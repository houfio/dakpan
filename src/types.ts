import { ComponentType, FunctionComponent } from 'react';

export type Dakpan<S, A extends Actions<S>> = [FunctionComponent, DakpanHook<S, A>, DakpanHoc<S, A>];

export type DakpanHook<S, A extends Actions<S>> = () => [S, MappedActions<S, A>];

export type DakpanHoc<S, A extends Actions<S>> = <H>(
  map: (state: S, actions: MappedActions<S, A>) => H
) => <P>(
  component: ComponentType<P & H>
) => ComponentType<P>;

export type DakpanContext<S, A extends Actions<S>> = undefined | {
  state: S,
  actions: MappedActions<S, A>
};

export type Actions<S> = {
  [action: string]: (...args: any[]) => (state: S) => S | undefined | Promise<S | undefined>
};

export type MappedActions<S, A extends Actions<S>> = {
  [K in keyof A]: ((...args: Parameters<A[K]>) => Promise<void>) & {
    c: (...args: Parameters<A[K]>) => () => Promise<void>
  }
};

export type InitialState<S> = S | (() => S);
