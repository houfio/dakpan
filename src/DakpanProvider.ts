import { createElement, ReactElement } from 'react';

import { DakpanMultiProviderProps } from './types';

export const DakpanProvider = ({ provide, children }: DakpanMultiProviderProps) => (
  provide.reduce((previous, current) => createElement(current, { children: previous }), children as ReactElement<any>)
);
