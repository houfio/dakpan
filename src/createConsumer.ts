import { ComponentClass, Consumer, ConsumerProps, createElement, ReactElement, StatelessComponent } from 'react';

import { Actions, DakpanConsumerProps, MappedActions } from './types';

export const createConsumer = <S, A extends Actions<S>>(
  Consumer: Consumer<S>,
  actions: MappedActions<S, A>
) => ({ children }: DakpanConsumerProps<S, A>) => createElement(Consumer, {
  children: (value: S) => children(value, actions)
});
