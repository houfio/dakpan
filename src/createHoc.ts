import { createElement, FunctionComponent } from 'react';

import { Actions, DakpanHoc, DakpanHook } from './types';

export const createHoc = <S, A extends Actions<S>>(
  hook: DakpanHook<S, A>
): DakpanHoc<S, A> => (map) => (component) => {
  const hoc: FunctionComponent<any> = (props) => {
    const [state, actions] = hook();

    return createElement(component, {
      ...props,
      ...map(state, actions)
    });
  };

  hoc.displayName = `WithDakpan(${component.displayName || component.name || 'Unknown'})`;

  return hoc;
};
