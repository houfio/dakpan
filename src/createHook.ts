import { Context, useContext } from 'react';

import { Actions, DakpanContext, DakpanHook } from './types';

export function createHook<S, A extends Actions<S>>(context: Context<DakpanContext<S, A>>): DakpanHook<S, A> {
  return () => {
    const state = useContext(context);

    if (!state) {
      throw Error('Provider not mounted');
    }

    return [
      state.state,
      state.actions
    ];
  };
}
