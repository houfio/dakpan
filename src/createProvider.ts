import { Context, createElement, FunctionComponent, useEffect, useMemo, useRef, useState } from 'react';

import { Actions, DakpanContext, MappedActions } from './types';

export function createProvider<S, A extends Actions<S>>(
  context: Context<DakpanContext<S, A>>,
  actions: A,
  state?: S
): FunctionComponent<any> {
  return ({ children, value, reinitialize = false }) => {
    const [, update] = useState({});
    const initialRef = useRef(value);
    const stateRef = useRef(state || value);
    const mapped = useMemo(() => Object.keys(actions).reduce(
      (previous, current) => {
        const execute = (...args: unknown[]) => () => Promise.resolve(actions[current](...args)(stateRef.current))
          .then((next) => {
            if (next) {
              stateRef.current = next;
              update({});
            }
          });

        const action = (...args: unknown[]) => execute(...args)();
        action.c = execute;

        return {
          ...previous,
          [current]: action
        };
      },
      {}
    ) as MappedActions<S, A>, [actions, stateRef]);

    useEffect(() => {
      if (reinitialize && value !== initialRef.current) {
        stateRef.current = initialRef.current = value;
        update({});
      }
    }, [value, reinitialize, initialRef]);

    return createElement(context.Provider, {
      value: {
        state: stateRef.current,
        actions: mapped
      }
    }, children);
  };
}
