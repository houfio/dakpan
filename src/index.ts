export { createDakpan } from './createDakpan';
export { DakpanProvider } from './DakpanProvider';
