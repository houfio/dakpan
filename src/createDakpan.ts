import {
  Component,
  ComponentClass,
  ConsumerProps,
  createContext,
  ProviderProps,
  ReactElement,
  ReactNodeArray,
  ReactPortal,
  StatelessComponent
} from 'react';

import { createConsumer } from './createConsumer';
import { createProvider } from './createProvider';
import { Actions, DakpanProviderProps, GetState, MappedActions, SetState } from './types';
import { withDakpan } from './withDakpan';

export const createDakpan = <S>(initialState: S) => <A extends Actions<S>>(actions: A) => {
  let getState: GetState<S> | undefined;
  let setState: SetState<S, DakpanProviderProps> | undefined;

  const mergeAction = (result: Partial<S>) => {
    if (!getState || !setState) {
      throw new Error('Provider unmounted before action could complete');
    } else if (typeof result !== 'object') {
      throw new Error('Actions may only return objects');
    }

    const newState = {
      ...getState() as any,
      ...result as any
    };

    setState(newState);

    return newState;
  };

  const { Provider, Consumer } = createContext<S>(initialState);
  const mappedActions = Object.keys(actions).reduce(
    (result, key) => {
      const e = (...args: any[]) => () => {
        const action = actions[key];

        if (!getState || !setState) {
          throw new Error('You may not dispatch an action before its provider is mounted');
        }

        const state = getState();
        const result = action(...args)(state);

        if (result instanceof Promise) {
          return result.then(mergeAction);
        }

        return mergeAction(result);
      };

      const action: any = (...args: any[]) => e(...args)();
      action.e = e;

      return {
        ...result,
        [key]: action
      };
    },
    {}
  ) as MappedActions<S, A>;

  const provider = createProvider(Provider, initialState, (get, set) => {
    getState = get;
    setState = set;
  });
  const consumer = createConsumer(Consumer, mappedActions);

  return {
    Provider: provider,
    Consumer: consumer,
    actions: mappedActions,
    withDakpan: withDakpan(consumer)
  };
};
