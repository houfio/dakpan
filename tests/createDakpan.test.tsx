import * as React from 'react';
import { Component } from 'react';
import { fireEvent, render } from 'react-testing-library';

import { createDakpan } from '../src';

const createMockDakpan = () => {
  const mockUpdate = jest.fn((current: number) => new Promise<number>((resolve) => resolve(current + 1)));
  const mockAction = jest.fn();

  return {
    ...createDakpan({
      hello: 'world',
      test: 0
    })({
      append: (append: string) => ({ hello }) => ({
        hello: `${hello}${append}`
      }),
      increment: () => async ({ test }) => ({
        test: await mockUpdate(test)
      }),
      mock: (test: string) => () => {
        mockAction(test);

        return {};
      },
      error: () => () => 'error' as any
    }),
    mockUpdate,
    mockAction
  };
};

it('should initialize the dakpan', () => {
  const { Provider, Consumer, actions } = createMockDakpan();

  expect(Provider).toBeDefined();
  expect(Consumer).toBeDefined();
  expect(actions).toBeDefined();
  expect(actions.append).toBeDefined();
});

it('should throw when an action is dispatched before its provider is mounted', () => {
  const { actions } = createMockDakpan();

  expect(actions.append.e('!')).toThrow('You may not dispatch an action before its provider is mounted');
});

it('should update the store', () => {
  const { Provider, Consumer, actions } = createMockDakpan();

  const { asFragment } = render(
    <Provider>
      <Consumer>
        {({ hello }) => hello}
      </Consumer>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();

  actions.append('!');

  expect(asFragment()).toMatchSnapshot();
});

it('should update the store asynchronously', async () => {
  const { Provider, Consumer, actions, mockUpdate } = createMockDakpan();

  const { asFragment } = render(
    <Provider>
      <Consumer>
        {({ test }) => test}
      </Consumer>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();

  await actions.increment();

  expect(mockUpdate).toBeCalledWith(0);
  expect(asFragment()).toMatchSnapshot();
});

it('should update the store as event handler', () => {
  const { Provider, Consumer, actions } = createMockDakpan();

  const { asFragment, getByTestId } = render(
    <Provider>
      <Consumer>
        {({ hello }) => (
          <>
            <span>{hello}</span>
            <button data-testid="button" onClick={actions.append.e('!')}/>
          </>
        )}
      </Consumer>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();

  const button = getByTestId('button');
  fireEvent.click(button);

  expect(asFragment()).toMatchSnapshot();
});

it('should call an action with the correct parameters', () => {
  const { Provider, actions, mockAction } = createMockDakpan();

  render(
    <Provider>
      <div/>
    </Provider>
  );

  actions.mock('hello!');
  const e = actions.mock.e('hello!!');

  expect(mockAction).toBeCalledWith('hello!');

  e();

  expect(mockAction).toBeCalledWith('hello!!');
});

it('should throw when the action doesn\'t return an object', () => {
  const { Provider, actions } = createMockDakpan();

  render(
    <Provider>
      <div/>
    </Provider>
  );

  expect(actions.error.e()).toThrow('Actions may only return objects');
});

it('should pass the actions to the consumer', () => {
  const { Provider, Consumer } = createMockDakpan();

  const { asFragment, getByTestId } = render(
    <Provider>
      <Consumer>
        {({ hello }, { append }) => (
          <>
            <span>{hello}</span>
            <button data-testid="button" onClick={append.e('!')}/>
          </>
        )}
      </Consumer>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();

  const button = getByTestId('button');
  fireEvent.click(button);

  expect(asFragment()).toMatchSnapshot();
});

it('should return the updated state', async () => {
  const { Provider, actions } = createMockDakpan();

  render(
    <Provider>
      <div/>
    </Provider>
  );

  expect(actions.append('!')).toMatchSnapshot();
  expect(await actions.increment()).toMatchSnapshot();
});

it('should pass the state to a stateless component using a hoc', () => {
  const { Provider, withConsumer } = createMockDakpan();

  type Test = {
    test: string
  };

  const Component = withConsumer(({ hello }, { append }) => ({
    hello,
    append
  }))<Test>(({ test, hello, append }) => (
    <>
      <span>{test}</span>
      <span>{hello}</span>
      <button onClick={append.e('!')}/>
    </>
  ));

  const { asFragment } = render(
    <Provider>
      <Component test="prop"/>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});

it('should pass the state to a stateful component using a hoc', () => {
  const { Provider, actions, withConsumer } = createMockDakpan();

  type Props = {
    test: string,
    hello: string,
    append: typeof actions.append
  };

  const TestComponent = withConsumer(({ hello }, { append }) => ({
    hello,
    append
  }))(class extends Component<Props> {
    public componentDidMount() {
      const { append } = this.props;

      append('!');
    }

    public render() {
      const { test, hello } = this.props;

      return (
        <>
          <span>{test}</span>
          <span>{hello}</span>
        </>
      );
    }
  });

  const { asFragment } = render(
    <Provider>
      <TestComponent test="prop"/>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
