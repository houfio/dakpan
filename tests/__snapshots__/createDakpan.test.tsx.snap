// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass the actions to the consumer 1`] = `
<DocumentFragment>
  <span>
    world
  </span>
  <button
    data-testid="button"
  />
</DocumentFragment>
`;

exports[`should pass the actions to the consumer 2`] = `
<DocumentFragment>
  <span>
    world!
  </span>
  <button
    data-testid="button"
  />
</DocumentFragment>
`;

exports[`should pass the state to a stateful component using a hoc 1`] = `
<DocumentFragment>
  <span>
    prop
  </span>
  <span>
    world!
  </span>
</DocumentFragment>
`;

exports[`should pass the state to a stateless component using a hoc 1`] = `
<DocumentFragment>
  <span>
    prop
  </span>
  <span>
    world
  </span>
  <button />
</DocumentFragment>
`;

exports[`should return the updated state 1`] = `
Object {
  "hello": "world!",
  "test": 0,
}
`;

exports[`should return the updated state 2`] = `
Object {
  "hello": "world!",
  "test": 1,
}
`;

exports[`should update the store 1`] = `
<DocumentFragment>
  world
</DocumentFragment>
`;

exports[`should update the store 2`] = `
<DocumentFragment>
  world!
</DocumentFragment>
`;

exports[`should update the store as event handler 1`] = `
<DocumentFragment>
  <span>
    world
  </span>
  <button
    data-testid="button"
  />
</DocumentFragment>
`;

exports[`should update the store as event handler 2`] = `
<DocumentFragment>
  <span>
    world!
  </span>
  <button
    data-testid="button"
  />
</DocumentFragment>
`;

exports[`should update the store asynchronously 1`] = `
<DocumentFragment>
  0
</DocumentFragment>
`;

exports[`should update the store asynchronously 2`] = `
<DocumentFragment>
  1
</DocumentFragment>
`;
